open_lm/main.py

```diff
@@ -467,17 +469,49 @@ def main(args):
     if args.train_data or (args.dataset_metadata is not None):
         named_parameters = list(model.named_parameters())
         no_decay_params = []  # to be potentially used later
-        params = [p for n, p in named_parameters if p.requires_grad]
 
-        optimizer = optim.AdamW(
-            [
-                {"params": no_decay_params, "weight_decay": 0.0},
-                {"params": params, "weight_decay": args.wd},
-            ],
-            lr=args.lr,
-            betas=(args.beta1, args.beta2),
-            eps=args.eps,
-        )
+        if args.mup_base_fan_in:
+            linear_modules = ["in_proj", "out_proj", "w12", "w3"]
+            mup_scale = args.mup_base_fan_in / model.params.dim
+            mup_lr = args.lr * mup_scale
+
+            mup_wd = args.wd
+            if args.mup_scale_wd:
+                # ensure that wd is still decoupled from lr even when the lr changes
+                mup_wd /= mup_scale
+
+            # modify all linear modules
+            linear_params = []
+            other_params = []
+            for n, p in named_parameters:
+                if p.requires_grad:
+                    if any(m in n for m in linear_modules):
+                        linear_params.append(p)
+                    else:
+                        other_params.append(p)
+
+            optimizer = optim.AdamW(
+                [
+                    {"params": no_decay_params, "lr": args.lr, "weight_decay": 0.0},
+                    {"params": other_params, "lr": args.lr, "weight_decay": args.wd},
+                    {"params": linear_params, "lr": mup_lr, "weight_decay": mup_wd},
```

Comment:
you'll have to special case w3 here

Reply:
now i am looping over all linear layers to get the specific fan-in from the weight shapes and creating param groups accordingly